Sources/Plasma/FeatureLib/pfPatcher/pfPatcher.cpp

```diff
@@ -705,10 +721,17 @@ void pfPatcher::OnSelfPatch(FileDownloadFunc cb)
 
 // ===================================================
 
-void pfPatcher::RequestGameCode()
+void pfPatcher::RequestGameCode(bool python, bool sdl)
 {
+    fWorker->fWantPython = python;
+    fWorker->fWantSDL = sdl;
+
     hsLockGuard(fWorker->fRequestMut);
-    fWorker->fRequests.emplace_back("SecurePreloader", pfPatcherWorker::Request::kSecurePreloader);
+    if (NetCliFileQueryConnected()) {
+        fWorker->fRequests.emplace_back("SecurePreloader", pfPatcherWorker::Request::kSecurePreloader);
+    } else {
+        fWorker->EnqueuePreloaderLists();
+    }
```

Comment:
Regardless of how the condition is spelled, I'm still unsure if this behaves as intended in some corner cases.

Specifically: if `/LocalData` is specified, but *not* `/LocalSDL`, the intended behavior is that .sdl files are fetched from the server and everything else is taken from the local files (without patching) - correct?

This code works that way if the .sdl files are available from the auth server/secure preloader, but *not* if the .sdl files are only served from the file server. Currently, enabling `/LocalData` makes the client never connect to the file server, even if one exists - so in that case, this code will only check the auth server and not the file server for .sdl files. The fix would be to change all the `if (!gDataServerLocal)` checks to `if (!gDataServerLocal && !gSDLLocal)` I think.

This *might* be a purely theoretical problem - I'm not sure. All I know is that DIRTSAND-based servers usually provide the .sdl files via the file server, but I don't know if that also means that they're *not* available from the auth server anymore.

In any case, it's not a big concern. This fix is primarily meant for Cyan's shard, where .sdl files are available from the auth server, so the code will work correctly there.

Reply:
> Specifically: if /LocalData is specified, but not /LocalSDL, the intended behavior is that .sdl files are fetched from the server and everything else is taken from the local files (without patching) - correct?

That's the way it's coded, yes. It could make more sense, in some ways, to only allow `/LocalSDL` if `/LocalData` is specified because using the shard's data server but your own bespoke SDL seems a little strange.

> This code works that way if the .sdl files are available from the auth server/secure preloader, but not if the .sdl files are only served from the file server. Currently, enabling /LocalData makes the client never connect to the file server, even if one exists - so in that case, this code will only check the auth server and not the file server for .sdl files. The fix would be to change all the conditions for connecting to the file server from if (!gDataServerLocal) to if (!gDataServerLocal && !gSDLLocal) I think.
> 
> This might be a purely theoretical problem - I'm not sure. All I know is that DIRTSAND-based servers usually provide the .sdl files via the file server, but I don't know if that also means that they're not available from the auth server anymore.
>
> In any case, it's not a big concern. This fix is primarily meant for Cyan's shard, where .sdl files are available from the auth server, so the code will work correctly there.

Yeah, we never actually wrote down any sort of guidelines about how Python and SDL should be served. In my mind, you *really* want to serve both Python and SDL from both locations - auth and file server. UruManifest will deploy Python and SDL for both auth and file servers, regardless of server software. To be honest, though, once you start playing around with `/LocalData` and especially `/LocalSDL`, you're explicitly turning off the safety valves and saying, "Yeah, I know what I'm doing. If things go badly, it's my fault." This PR is mostly to prevent bad things from happening to other people playing alongside you with some fringe benefits for people using H'uru clients on MOULa.

Reply:
> In my mind, you _really_ want to serve both Python and SDL from both locations - auth and file server.

Okay - if that's the preferred/common practice, then this code is fine.

(I made a small logic error above btw: if we were to change the file server connect conditions, they should become `if (!gDataServerLocal || !gSDLLocal)`, with a `||` rather than `&&` - but that doesn't really matter now.)